message/message-factory.rst

```diff
@@ -1,57 +1,124 @@
-
-.. _stream-factory:
+Factories
+---------
 
 The `php-http/message-factory` package defines interfaces for PSR-7 factories including:
 
-- ``MessageFactory``
-- ``ServerRequestFactory`` - WIP (PRs welcome)
+- ``RequestFactory``
+- ``ResponseFactory``
+- ``MessageFactory`` (combination of request and response factories)
 - ``StreamFactory``
-- ``UploadedFileFactory`` - WIP (PRs welcome)
 - ``UriFactory``
 
-Implementation for the interfaces above for `Diactoros`_ and `Guzzle PSR-7`_ and `Slim Framework`_ can be found in ``php-http/message``.
-
-.. code:: php
+Implementations of the interfaces above for `Diactoros`_, `Guzzle PSR-7`_ and the `Slim Framework`_ can be found in ``php-http/message``.
 
-    // Create a PSR-7 request
-    $factory = new Http\Message\MessageFactory\DiactorosMessageFactory();
-    $request = $factory->createRequest('GET', 'http://example.com');
-
-    // Create a PSR-7 stream
-    $factory = new Http\Message\StreamFactory\DiactorosStreamFactory();
-    $stream = $factory->createStream('stream content');
-
-You could also use :doc:`/discovery` to find an installed factory automatically.
-
-.. code:: php
-
-    // Create a PSR-7 request
-    $factory = MessageFactoryDiscovery::find();
-    $request = $factory->createRequest('GET', 'http://example.com');
+Usage
+-----
 
+Instantiate the factories in your bootstrap code or use discovery for them.
+Inject the factories into the rest of your code to limit the implementation
+choice to the bootstrapping code::
+
+    // ApiClient.php
+
+    use Http\Message\RequestFactory;
+    use Http\Message\StreamFactory;
+    use Http\Message\UriFactory;
+
+    class ApiClient
+    {
+        /**
+         * @var RequestFactory
+         */
+        private $requestFactory;
+
+        /**
+         * @var StreamFactory
+         */
+        private $streamFactory;
+
+        /**
+         * @var UriFactory
+         */
+        private $uriFactory;
+
+        public function __construct(
+            RequestFactory $requestFactory,
+            StreamFactory $streamFactory,
+            UriFactory $uriFactory
+        ) {
+            $this->requestFactory = $requestFactory;
+            $this->streamFactory = $streamFactory;
+            $this->uriFactory = $uriFactory;
+        }
+
+        public function doStuff()
+        {
+            $request = $this->requestFactory->createRequest('GET', 'http://httplug.io');
+            $stream = $this->streamFactory->createStream('stream content');
+            $uri = $this->UriFactory->createUri('http://httplug.io');
+            ...
+        }
+    }
+
+The bootstrapping code could look like this::
+
+    // bootstrap.php
+    use Http\Message\MessageFactory\DiactorosMessageFactory;
+    use Http\Message\StreamFactory\DiactorosStreamFactory;
+    use Http\Message\UriFactory\DiactorosUriFactory;
+
+    $apiClient = new ApiClient(
+        new DiactorosMessageFactory(),
+        new DiactorosStreamFactory(),
+        new DiactorosUriFactory()
```

Comment:
You would rarely use the URI Factory since it is required by the message factory. 

I suggest removing it.